onnx/optimizer/passes/fuse_matmul_add_bias_into_gemm.h

```diff
@@ -0,0 +1,97 @@
+// ATTENTION: The code in this file is highly EXPERIMENTAL.
+// Adventurous users should note that the APIs will probably change.
+
+#pragma once
+
+// Before:
+//   Z = MatMul(X, Y)
+//   A = Z + Bias
+// After:
+//   A = Gemm(X, Y, Bias)
+//
+// the pass can handle the case when:
+//   case 1: Bias is 1D tensor and Bias.dim[0] == Z.dim[1]
+//   case 2: Bias is 2D tensor and Bias.dim[0] == Z.dim[0] or 1
+//           and Bias.dim[1] = Z.dim[1]
+
+#include <numeric>
+
+#include "onnx/common/assertions.h"
+#include "onnx/optimizer/pass.h"
+
+namespace ONNX_NAMESPACE {
+namespace optimization {
+
+struct FuseMatMulAddBiasIntoGemm final : public PredicateBasedPass {
+  explicit FuseMatMulAddBiasIntoGemm()
+      : PredicateBasedPass(
+            PassType::Fuse,
+            PassEfficiency::Complete,
+            PassOptimizationType::Compute) {}
+  std::string getPassName() const override {
+    return "fuse_matmul_add_bias_into_gemm";
+  }
+  bool patternMatchPredicate(Node* node) override {
+    return node->kind() == kAdd &&
+        node->inputs()[0]->node()->kind() == kMatMul;
+  }
+  bool runTransform(Node* n, Graph& graph, NodeDestroyType& destroy_current)
+      override {
+    // due to current broadcasting's constraint, MatMul has to be the first
+    // operand
+    destroy_current = NodeDestroyType::DestroyZero;
+    auto orig_matmul = n->inputs()[0];
+    auto orig_bias = n->inputs()[1];
+    // check if bias is Const or in graph's initializers
+    if (orig_bias->node()->kind() != kConstant &&
+        orig_bias->node()->kind() != kParam) {
+      return false;
+    }
+    // check if MatMul is only used by Add
+    if (orig_matmul->uses().size() > 1) {
+      return false;
+    }
+    auto matmul_shape = orig_matmul->node()->inputs()[1]->sizes();
+    auto matmul_dim = static_cast<int64_t>(matmul_shape.size());
+    int64_t matmul_M = -1;
+    // try to get feature M from matmul_shape
+    if (matmul_dim == 2 && matmul_shape[1].is_int) {
+      matmul_M = matmul_shape[1].dim;
+    } else {
+      return false;
+    }
+    // check if bias_shape is compatible
+    auto bias_shape = orig_bias->sizes();
+    auto bias_dim = static_cast<int64_t>(bias_shape.size());
+    int64_t bias_M = -1;
+    if (bias_dim == 1 && bias_shape[0].is_int) {
+      bias_M = bias_shape[0].dim;
+    } else if (bias_dim == 2 && bias_shape[1].is_int) {
```

Comment:
there is still a problem here. Since in GEMM we only broadcast the bias to the A*B. so we should check here whether bias's second dimension is 1 or the same as A*B.

Reply:
I had the stricter check in the previous version, but thought it was too much. I have reintroduced it now.